From 6c2aace8d6e6ea97d741645a66042e02bdbffa4b Mon Sep 17 00:00:00 2001
From: Matt Crees <mattc@stackhpc.com>
Date: Tue, 15 Nov 2022 13:46:53 +0000
Subject: [PATCH] Integrate oslo-config-validator

Regularly, we experience issues in Kolla Ansible deployments because we
use wrong options in OpenStack configuration files. This is because
OpenStack services ignore unknown options. We also need to keep on top
of deprecated options that may be removed in the future. Integrating
oslo-config-validator into Kolla Ansible will greatly help.

Adds a shared role to run oslo-config-validator on each service. Takes
into account that services have multiple containers, and these may also
use multiple config files. Service roles are extended to use this shared
role. Executed with the new command ``kolla-ansible validate-config``.

Change-Id: Ic10b410fc115646d96d2ce39d9618e7c46cb3fbc
---

diff --git a/ansible/roles/aodh/defaults/main.yml b/ansible/roles/aodh/defaults/main.yml
index 0f8a29d71..19f56f09e 100644
--- a/ansible/roles/aodh/defaults/main.yml
+++ b/ansible/roles/aodh/defaults/main.yml
@@ -55,6 +55,12 @@ aodh_services:
 # change this option.
 aodh_evaluation_interval: 300
 
+####################
+# Config Validate
+####################
+aodh_config_validation:
+  - generator: "/aodh/aodh/cmd/aodh-config-generator.conf"
+    config: "/etc/aodh/aodh.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/barbican/defaults/main.yml b/ansible/roles/barbican/defaults/main.yml
index 0f7778175..f8629a5a6 100644
--- a/ansible/roles/barbican/defaults/main.yml
+++ b/ansible/roles/barbican/defaults/main.yml
@@ -40,6 +40,12 @@ barbican_services:
     dimensions: "{{ barbican_worker_dimensions }}"
     healthcheck: "{{ barbican_worker_healthcheck }}"
 
+####################
+# Config Validate
+####################
+barbican_config_validation:
+  - generator: "/barbican/etc/oslo-config-generator/barbican.conf"
+    config: "/etc/barbican/barbican.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/blazar/defaults/main.yml b/ansible/roles/blazar/defaults/main.yml
index 36d9d24a1..bdcf00bc5 100644
--- a/ansible/roles/blazar/defaults/main.yml
+++ b/ansible/roles/blazar/defaults/main.yml
@@ -34,6 +34,13 @@ blazar_services:
 ####################
 blazar_aggregate_pool_name: "freepool"
 
+####################
+# Config Validate
+####################
+blazar_config_validation:
+  - generator: "/blazar/etc/blazar/blazar-config-generator.conf"
+    config: "/etc/blazar/blazar.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/ceilometer/defaults/main.yml b/ansible/roles/ceilometer/defaults/main.yml
index 7cbcd5674..d9024926e 100644
--- a/ansible/roles/ceilometer/defaults/main.yml
+++ b/ansible/roles/ceilometer/defaults/main.yml
@@ -34,6 +34,12 @@ ceilometer_services:
     dimensions: "{{ ceilometer_ipmi_dimensions }}"
     healthcheck: "{{ ceilometer_ipmi_healthcheck }}"
 
+####################
+# Config Validate
+####################
+ceilometer_config_validation:
+  - generator: "/ceilometer/etc/ceilometer/ceilometer-config-generator.conf"
+    config: "/etc/ceilometer/ceilometer.conf"
 
 ####################
 # Docker
diff --git a/ansible/roles/cinder/defaults/main.yml b/ansible/roles/cinder/defaults/main.yml
index ab764edcc..f33e517d0 100644
--- a/ansible/roles/cinder/defaults/main.yml
+++ b/ansible/roles/cinder/defaults/main.yml
@@ -52,6 +52,13 @@ cinder_services:
     dimensions: "{{ cinder_backup_dimensions }}"
     healthcheck: "{{ cinder_backup_healthcheck }}"
 
+####################
+# Config Validate
+####################
+cinder_config_validation:
+  - generator: "/cinder/tools/config/cinder-config-generator.conf"
+    config: "/etc/cinder/cinder.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/cloudkitty/defaults/main.yml b/ansible/roles/cloudkitty/defaults/main.yml
index dedf4161f..e10872655 100644
--- a/ansible/roles/cloudkitty/defaults/main.yml
+++ b/ansible/roles/cloudkitty/defaults/main.yml
@@ -28,6 +28,12 @@ cloudkitty_services:
     dimensions: "{{ cloudkitty_processor_dimensions }}"
     healthcheck: "{{ cloudkitty_processor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+cloudkitty_config_validation:
+  - generator: "/cloudkitty/etc/oslo-config-generator/cloudkitty.conf"
+    config: "/etc/cloudkitty/cloudkitty.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/cyborg/defaults/main.yml b/ansible/roles/cyborg/defaults/main.yml
index eb92290b8..69ca507f6 100644
--- a/ansible/roles/cyborg/defaults/main.yml
+++ b/ansible/roles/cyborg/defaults/main.yml
@@ -26,6 +26,13 @@ cyborg_services:
     dimensions: "{{ cyborg_conductor_dimensions }}"
     healthcheck: "{{ cyborg_conductor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+cyborg_config_validation:
+  - generator: "/cyborg/tools/config/cyborg-config-generator.conf"
+    config: "/etc/cyborg/cyborg.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/designate/defaults/main.yml b/ansible/roles/designate/defaults/main.yml
index 9619a1689..926a358b4 100644
--- a/ansible/roles/designate/defaults/main.yml
+++ b/ansible/roles/designate/defaults/main.yml
@@ -70,6 +70,12 @@ designate_services:
     dimensions: "{{ designate_sink_dimensions }}"
     healthcheck: "{{ designate_sink_healthcheck }}"
 
+####################
+# Config Validate
+####################
+designate_config_validation:
+  - generator: "/designate/etc/designate/designate-config-generator.conf"
+    config: "/etc/designate/designate.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/freezer/defaults/main.yml b/ansible/roles/freezer/defaults/main.yml
index 1956c1e26..716fdc432 100644
--- a/ansible/roles/freezer/defaults/main.yml
+++ b/ansible/roles/freezer/defaults/main.yml
@@ -26,6 +26,13 @@ freezer_services:
     volumes: "{{ freezer_scheduler_default_volumes + freezer_scheduler_extra_volumes }}"
     dimensions: "{{ freezer_scheduler_dimensions }}"
 
+####################
+# Config Validate
+####################
+freezer_config_validation:
+  - generator: "/freezer/etc/config-generator.conf"
+    config: "/etc/freezer/freezer.conf"
+
 ####################
 ## Database
 #####################
diff --git a/ansible/roles/glance/defaults/main.yml b/ansible/roles/glance/defaults/main.yml
index 5ca70d9c7..114cce55e 100644
--- a/ansible/roles/glance/defaults/main.yml
+++ b/ansible/roles/glance/defaults/main.yml
@@ -65,6 +65,19 @@ glance_services:
         custom_member_list: "{{ haproxy_tls_members.split(';') }}"
         tls_backend: "yes"
 
+####################
+# Config Validate
+####################
+glance_config_validation:
+  - generator: "/glance/etc/oslo-config-generator/glance-api.conf"
+    config: "/etc/glance/glance-api.conf"
+  - generator: "/glance/etc/oslo-config-generator/glance-cache.conf"
+    config: "/etc/glance/glance-cache.conf"
+  - generator: "/glance/etc/oslo-config-generator/glance-manage.conf"
+    config: "/etc/glance/glance-manage.conf"
+  - generator: "/glance/etc/oslo-config-generator/glance-scrubber.conf"
+    config: "/etc/glance/glance-scrubber.conf"
+
 ####################
 # HAProxy
 ####################
diff --git a/ansible/roles/gnocchi/defaults/main.yml b/ansible/roles/gnocchi/defaults/main.yml
index e15582f50..dbb5ef5e3 100644
--- a/ansible/roles/gnocchi/defaults/main.yml
+++ b/ansible/roles/gnocchi/defaults/main.yml
@@ -36,6 +36,12 @@ gnocchi_services:
     dimensions: "{{ gnocchi_statsd_dimensions }}"
     healthcheck: "{{ gnocchi_statsd_healthcheck }}"
 
+####################
+# Config Validate
+####################
+gnocchi_config_validation:
+  - generator: "/gnocchi/gnocchi/gnocchi-config-generator.conf"
+    config: "/etc/gnocchi/gnocchi.conf"
 
 ####################
 # Swift
diff --git a/ansible/roles/heat/defaults/main.yml b/ansible/roles/heat/defaults/main.yml
index 37a4622f6..d4b630df1 100644
--- a/ansible/roles/heat/defaults/main.yml
+++ b/ansible/roles/heat/defaults/main.yml
@@ -55,6 +55,13 @@ heat_services:
     dimensions: "{{ heat_engine_dimensions }}"
     healthcheck: "{{ heat_engine_healthcheck }}"
 
+####################
+# Config Validate
+####################
+heat_config_validation:
+  - generator: "/heat/config-generator.conf"
+    config: "/etc/heat/heat.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/ironic/defaults/main.yml b/ansible/roles/ironic/defaults/main.yml
index ad735938b..7684cf97a 100644
--- a/ansible/roles/ironic/defaults/main.yml
+++ b/ansible/roles/ironic/defaults/main.yml
@@ -84,6 +84,12 @@ ironic_services:
     volumes: "{{ ironic_dnsmasq_default_volumes + ironic_dnsmasq_extra_volumes }}"
     dimensions: "{{ ironic_dnsmasq_dimensions }}"
 
+####################
+# Config Validate
+####################
+ironic_config_validation:
+  - generator: "/ironic/tools/config/ironic-config-generator.conf"
+    config: "/etc/ironic/ironic.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/keystone/defaults/main.yml b/ansible/roles/keystone/defaults/main.yml
index d0710e308..94de04d97 100644
--- a/ansible/roles/keystone/defaults/main.yml
+++ b/ansible/roles/keystone/defaults/main.yml
@@ -61,6 +61,13 @@ keystone_services:
     dimensions: "{{ keystone_fernet_dimensions }}"
     healthcheck: "{{ keystone_fernet_healthcheck }}"
 
+####################
+# Config Validate
+####################
+keystone_config_validation:
+  - generator: "/keystone/config-generator/keystone.conf"
+    config: "/etc/keystone/keystone.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/kuryr/defaults/main.yml b/ansible/roles/kuryr/defaults/main.yml
index 92c60fc5b..0a2cbfa62 100644
--- a/ansible/roles/kuryr/defaults/main.yml
+++ b/ansible/roles/kuryr/defaults/main.yml
@@ -20,6 +20,12 @@ kuryr_services:
     dimensions: "{{ kuryr_dimensions }}"
     healthcheck: "{{ kuryr_healthcheck }}"
 
+####################
+# Config Validate
+####################
+kuryr_config_validation:
+  - generator: "/kuryr/etc/kuryr-config-generator.conf"
+    config: "/etc/kuryr/kuryr.conf"
 
 ####################
 # Docker
diff --git a/ansible/roles/magnum/defaults/main.yml b/ansible/roles/magnum/defaults/main.yml
index 1a051095d..c929842e5 100644
--- a/ansible/roles/magnum/defaults/main.yml
+++ b/ansible/roles/magnum/defaults/main.yml
@@ -31,6 +31,12 @@ magnum_services:
     dimensions: "{{ magnum_conductor_dimensions }}"
     healthcheck: "{{ magnum_conductor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+magnum_config_validation:
+  - generator: "/magnum/etc/magnum/magnum-config-generator.conf"
+    config: "/etc/magnum/magnum.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/manila/defaults/main.yml b/ansible/roles/manila/defaults/main.yml
index 895290a32..8f5a13f1d 100644
--- a/ansible/roles/manila/defaults/main.yml
+++ b/ansible/roles/manila/defaults/main.yml
@@ -46,6 +46,12 @@ manila_services:
     dimensions: "{{ manila_data_dimensions }}"
     healthcheck: "{{ manila_data_healthcheck }}"
 
+####################
+# Config Validate
+####################
+manila_config_validation:
+  - generator: "/manila/etc/oslo-config-generator/manila.conf"
+    config: "/etc/manila/manila.conf"
 
 #####################
 ## Database
diff --git a/ansible/roles/masakari/defaults/main.yml b/ansible/roles/masakari/defaults/main.yml
index d9bc87a6f..e581e2ac4 100644
--- a/ansible/roles/masakari/defaults/main.yml
+++ b/ansible/roles/masakari/defaults/main.yml
@@ -42,6 +42,12 @@ masakari_services:
     volumes: "{{ masakari_hostmonitor_default_volumes + masakari_hostmonitor_extra_volumes }}"
     dimensions: "{{ masakari_hostmonitor_dimensions }}"
 
+####################
+# Config Validate
+####################
+masakari_config_validation:
+  - generator: "/masakari/etc/masakari/masakari-config-generator.conf"
+    config: "/etc/masakari/masakari.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/mistral/defaults/main.yml b/ansible/roles/mistral/defaults/main.yml
index b00acd8da..3d1e2c64d 100644
--- a/ansible/roles/mistral/defaults/main.yml
+++ b/ansible/roles/mistral/defaults/main.yml
@@ -44,6 +44,12 @@ mistral_services:
     dimensions: "{{ mistral_executor_dimensions }}"
     healthcheck: "{{ mistral_executor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+mistral_config_validation:
+  - generator: "/mistral/tools/config/config-generator.mistral.conf"
+    config: "/etc/mistral/mistral.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/murano/defaults/main.yml b/ansible/roles/murano/defaults/main.yml
index e97ce75db..473a3d189 100644
--- a/ansible/roles/murano/defaults/main.yml
+++ b/ansible/roles/murano/defaults/main.yml
@@ -26,6 +26,12 @@ murano_services:
     volumes: "{{ murano_engine_default_volumes + murano_engine_extra_volumes }}"
     dimensions: "{{ murano_engine_dimensions }}"
 
+####################
+# Config Validate
+####################
+murano_config_validation:
+  - generator: "/murano/etc/oslo-config-generator/murano.conf"
+    config: "/etc/murano/murano.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/neutron/defaults/main.yml b/ansible/roles/neutron/defaults/main.yml
index 696011202..09f6f91bf 100644
--- a/ansible/roles/neutron/defaults/main.yml
+++ b/ansible/roles/neutron/defaults/main.yml
@@ -198,6 +198,33 @@ neutron_services:
         listen_port: "{{ neutron_server_listen_port }}"
         tls_backend: "yes"
 
+####################
+# Config Validate
+####################
+neutron_config_validation:
+  - generator: "/neutron/etc/oslo-config-generator/neutron.conf"
+    config: "/etc/neutron/neutron.conf"
+  - generator: "/neutron/etc/oslo-config-generator/neutron.conf"
+    config: "/etc/neutron/neutron_vpnaas.conf"
+  - generator: "/neutron/etc/oslo-config-generator/ml2_conf.ini"
+    config: "/etc/neutron/plugins/ml2/ml2_conf.ini"
+  - generator: "/neutron/etc/oslo-config-generator/openvswitch_agent.ini"
+    config: "/etc/neutron/plugins/ml2/openvswitch_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/metering_agent.ini"
+    config: "/etc/neutron/metering_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/neutron_ovn_metadata_agent.ini"
+    config: "/etc/neutron/neutron_ovn_metadata_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/metadata_agent.ini"
+    config: "/etc/neutron/metadata_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/sriov_agent.ini"
+    config: "/etc/neutron/plugins/ml2/sriov_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/l3_agent.ini"
+    config: "/etc/neutron/l3_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/dhcp_agent.ini"
+    config: "/etc/neutron/dhcp_agent.ini"
+  - generator: "/neutron/etc/oslo-config-generator/linuxbridge_agent.ini"
+    config: "/etc/neutron/plugins/ml2/linuxbridge_agent.ini"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/nova-cell/defaults/main.yml b/ansible/roles/nova-cell/defaults/main.yml
index a976c769e..0eead7ffb 100644
--- a/ansible/roles/nova-cell/defaults/main.yml
+++ b/ansible/roles/nova-cell/defaults/main.yml
@@ -71,6 +71,13 @@ nova_cell_services:
     dimensions: "{{ nova_compute_ironic_dimensions }}"
     healthcheck: "{{ nova_compute_ironic_healthcheck }}"
 
+####################
+# Config Validate
+####################
+nova_cell_config_validation:
+  - generator: "/nova/etc/nova/nova-config-generator.conf"
+    config: "/etc/nova/nova.conf"
+
 ####################
 # Ceph options
 ####################
diff --git a/ansible/roles/nova/defaults/main.yml b/ansible/roles/nova/defaults/main.yml
index b6a23e2f9..dad4ab678 100644
--- a/ansible/roles/nova/defaults/main.yml
+++ b/ansible/roles/nova/defaults/main.yml
@@ -55,6 +55,13 @@ nova_services:
     dimensions: "{{ nova_super_conductor_dimensions }}"
     healthcheck: "{{ nova_super_conductor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+nova_config_validation:
+  - generator: "/nova/etc/nova/nova-config-generator.conf"
+    config: "/etc/nova/nova.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/octavia/defaults/main.yml b/ansible/roles/octavia/defaults/main.yml
index ae428d4c7..49c502b43 100644
--- a/ansible/roles/octavia/defaults/main.yml
+++ b/ansible/roles/octavia/defaults/main.yml
@@ -62,6 +62,12 @@ octavia_required_roles:
   - load-balancer_admin
   - load-balancer_quota_admin
 
+####################
+# Config Validate
+####################
+octavia_config_validation:
+  - generator: "/octavia/etc/config/octavia-config-generator.conf"
+    config: "/etc/octavia/octavia.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/placement/defaults/main.yml b/ansible/roles/placement/defaults/main.yml
index a287454e6..51c78739b 100644
--- a/ansible/roles/placement/defaults/main.yml
+++ b/ansible/roles/placement/defaults/main.yml
@@ -24,6 +24,13 @@ placement_services:
         listen_port: "{{ placement_api_listen_port }}"
         tls_backend: "{{ placement_enable_tls_backend }}"
 
+####################
+# Config Validate
+####################
+placement_config_validation:
+  - generator: "/placement/etc/placement/config-generator.conf"
+    config: "/etc/placement/placement.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/sahara/defaults/main.yml b/ansible/roles/sahara/defaults/main.yml
index 7d746c23b..dadea0f29 100644
--- a/ansible/roles/sahara/defaults/main.yml
+++ b/ansible/roles/sahara/defaults/main.yml
@@ -29,6 +29,12 @@ sahara_services:
     dimensions: "{{ sahara_engine_dimensions }}"
     healthcheck: "{{ sahara_engine_healthcheck }}"
 
+####################
+# Config Validate
+####################
+sahara_config_validation:
+  - generator: "/sahara/tools/config/config-generator.sahara.conf"
+    config: "/etc/sahara/sahara.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/senlin/defaults/main.yml b/ansible/roles/senlin/defaults/main.yml
index ded678fb3..b1bfa1108 100644
--- a/ansible/roles/senlin/defaults/main.yml
+++ b/ansible/roles/senlin/defaults/main.yml
@@ -46,6 +46,13 @@ senlin_services:
     dimensions: "{{ senlin_health_manager_dimensions }}"
     healthcheck: "{{ senlin_health_manager_healthcheck }}"
 
+####################
+# Config Validate
+####################
+senlin_config_validation:
+  - generator: "/senlin/tools/config-generator.conf"
+    config: "/etc/senlin/senlin.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/solum/defaults/main.yml b/ansible/roles/solum/defaults/main.yml
index 766dea8ec..4af0633a6 100644
--- a/ansible/roles/solum/defaults/main.yml
+++ b/ansible/roles/solum/defaults/main.yml
@@ -58,6 +58,13 @@ solum_services:
     dimensions: "{{ solum_conductor_dimensions }}"
     healthcheck: "{{ solum_conductor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+solum_config_validation:
+  - generator: "/solum/etc/solum/config-generator.conf"
+    config: "/etc/solum/solum.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/tacker/defaults/main.yml b/ansible/roles/tacker/defaults/main.yml
index 8ec823378..daa9980b0 100644
--- a/ansible/roles/tacker/defaults/main.yml
+++ b/ansible/roles/tacker/defaults/main.yml
@@ -32,6 +32,13 @@ tacker_services:
     dimensions: "{{ tacker_conductor_dimensions }}"
     healthcheck: "{{ tacker_conductor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+tacker_config_validation:
+  - generator: "/tacker/etc/config-generator.conf"
+    config: "/etc/tacker/tacker.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/trove/defaults/main.yml b/ansible/roles/trove/defaults/main.yml
index 1971f4a36..455849ff3 100644
--- a/ansible/roles/trove/defaults/main.yml
+++ b/ansible/roles/trove/defaults/main.yml
@@ -36,6 +36,12 @@ trove_services:
     dimensions: "{{ trove_taskmanager_dimensions }}"
     healthcheck: "{{ trove_taskmanager_healthcheck }}"
 
+####################
+# Config Validate
+####################
+trove_config_validation:
+  - generator: "/trove/tools/trove-config-generator.conf"
+    config: "/etc/trove/trove.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/venus/defaults/main.yml b/ansible/roles/venus/defaults/main.yml
index a1a7021a1..d2d044f59 100644
--- a/ansible/roles/venus/defaults/main.yml
+++ b/ansible/roles/venus/defaults/main.yml
@@ -27,6 +27,13 @@ venus_services:
     volumes: "{{ venus_manager_default_volumes + venus_manager_extra_volumes }}"
     dimensions: "{{ venus_manager_dimensions }}"
 
+####################
+# Config Validate
+####################
+venus_config_validation:
+  - generator: "/venus/tools/config/venus-config-generator.conf"
+    config: "/etc/venus/venus.conf"
+
 ####################
 # Database
 ####################
diff --git a/ansible/roles/vitrage/defaults/main.yml b/ansible/roles/vitrage/defaults/main.yml
index bbe74ebf5..8e192d784 100644
--- a/ansible/roles/vitrage/defaults/main.yml
+++ b/ansible/roles/vitrage/defaults/main.yml
@@ -52,6 +52,13 @@ vitrage_services:
     dimensions: "{{ vitrage_persistor_dimensions }}"
     healthcheck: "{{ vitrage_persistor_healthcheck }}"
 
+####################
+# Config Validate
+####################
+vitrage_config_validation:
+  - generator: "/vitrage/etc/vitrage/vitrage-config-generator.conf"
+    config: "/etc/vitrage/vitrage.conf"
+
 ####################
 ## Database
 #####################
diff --git a/ansible/roles/watcher/defaults/main.yml b/ansible/roles/watcher/defaults/main.yml
index 35df766ec..03f344553 100644
--- a/ansible/roles/watcher/defaults/main.yml
+++ b/ansible/roles/watcher/defaults/main.yml
@@ -36,6 +36,12 @@ watcher_services:
     dimensions: "{{ watcher_engine_dimensions }}"
     healthcheck: "{{ watcher_engine_healthcheck }}"
 
+####################
+# Config Validate
+####################
+watcher_config_validation:
+  - generator: "/watcher/etc/watcher/oslo-config-generator/watcher.conf"
+    config: "/etc/watcher/watcher.conf"
 
 ####################
 # Database
diff --git a/ansible/roles/zun/defaults/main.yml b/ansible/roles/zun/defaults/main.yml
index edb727c9c..61719a179 100644
--- a/ansible/roles/zun/defaults/main.yml
+++ b/ansible/roles/zun/defaults/main.yml
@@ -57,6 +57,13 @@ zun_services:
     dimensions: "{{ zun_cni_daemon_dimensions }}"
     healthcheck: "{{ zun_cni_daemon_healthcheck }}"
 
+####################
+# Config Validate
+####################
+zun_config_validation:
+  - generator: "/zun/etc/zun/zun-config-generator.conf"
+    config: "/etc/zun/zun.conf"
+
 ####################
 ## Database
 ####################
diff --git a/tools/kolla-ansible b/tools/kolla-ansible
index 861faec74..3766db617 100755
--- a/tools/kolla-ansible
+++ b/tools/kolla-ansible
@@ -200,6 +200,7 @@ Commands:
     upgrade              Upgrades existing OpenStack Environment
     upgrade-bifrost      Upgrades an existing bifrost container
     genconfig            Generate configuration files for enabled OpenStack services
+    validate-config      Validate configuration files for enabled OpenStack services
     prune-images         Prune orphaned Kolla images
     nova-libvirt-cleanup Clean up disabled nova_libvirt containers
 EOF
@@ -246,6 +247,7 @@ octavia-certificates
 upgrade
 upgrade-bifrost
 genconfig
+validate-config
 prune-images
 nova-libvirt-cleanup
 EOF
@@ -535,6 +537,10 @@ EOF
         ACTION="Generate configuration files for enabled OpenStack services"
         EXTRA_OPTS="$EXTRA_OPTS -e kolla_action=config"
         ;;
+(validate-config)
+        ACTION="Validate configuration files for enabled OpenStack services"
+        EXTRA_OPTS="$EXTRA_OPTS -e kolla_action=config_validate"
+        ;;
 (prune-images)
         ACTION="Prune orphaned Kolla images"
         PLAYBOOK="${BASEDIR}/ansible/prune-images.yml"
